code/DynamicCache.php

```diff
@@ -397,14 +412,32 @@ public function run($url)
         // This is normally called in VersionedRequestFilter.
         Versioned::choose_site_stage();
 
+        // Get cache
```

Comment:
Given your example above, I would suggest to use the specific cookie to generate the additional cache key in updateCacheKeyFragments, and simply set this cookie to the appropriate variant in your PageController::init() method.

What I would suggest, however, is a minor change to DynamicCache, so that after handling a cache-hit the cache key is re-generated just before saving the cached result. That way your code could assign this key, and prepare the cached result for that key at the same time (subsequent requests will include the cookie, and thus his this new key).

Maybe this `post-request` key could also take the cookies / headers that are being saved, as well as the result? You could populate your secondary user-code level cache here.

Would this meet the use case you are trying to address? Ideally we could solve this problem without the extra complexity this PR is introducing; Sorry if that sounds overly critical. :)